### PR TITLE
Fix bounds of match type cases

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala [hunk 1]
```diff
@@ -401,15 +401,19 @@ trait TypeAssigner {
...
           }
         }
-        HKTypeLambda.fromParams(
-          params(new mutable.ListBuffer[TypeSymbol](), pat).toList,
-          defn.MatchCase(pat.tpe, body.tpe))
+        val params = getParams(new mutable.ListBuffer[TypeSymbol](), pat).toList
+        pat.tpe match
+          case AppliedType(tycon, args) =>
+            val tparams = tycon.typeParamSymbols
+            for param <- params do param.info = param.info.subst(tparams, args)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:414]: I don't know what this does exactly but you almost certainly shouldn't actually mutate types and instead should create a copy of the params with different types.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:414]: Gladly.  How do I do that?

Comment (smarter) [reply compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:414]: param.copy(info = ...)

Comment (smarter) [reply compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:414]: oh, but those are pattern-bound symbols? Then I'm not sure if they should be copied or not. If they could be created with the correct type that'd be much better.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:414]: I wish I could create them with the correct type, but I don't think I can, in an order independent way.  I iterate over them with their expected type, and then I need to substitute any references among those pattern-bound symbol types.  Unless I'm missing something?

Comment (smarter) [reply compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:414]: If you create them and then mutate their info in the same block of code, that would be fine.

**Review by smarter — COMMENTED** [review]
I don't have a full understanding of the match type logic but my gut reaction here is that this is a lot of code in TypeAssigner which is supposed to be mostly straightforward. Can we ensure these symbols have the correct info at the point they're created (both in typer and treeunpicker) instead?

**Comment by dwijnand** [comment]
I don't see how to do that.  The loop is in typing the arg of the AppliedTypeTree (typedAppliedTypeTree) and the symbols are created in typing the bind (typedBind).  In my example I need the symbol for A so I can use it as the upper bound of the symbol for X.  But there's no restriction on the definition order, so I could be typing the binding for X before I've even got a symbol for A.

So unless you can see another way to pipe and order things, I propose we use the substitute mechanism to post-process them.  There's seems to be a bunch of precedent for doing things very similarly to how I am, or at least to my eyes...

**Comment by dwijnand** [comment]
cc also @odersky, in case can recommend better ways or places to do this.

**Review by OlivierBlanvillain — APPROVED** [review]
LGTM! I'm happy that we fix this at the source, in TypeAssigner.

**Pull request merged**